pkg/cmd/delete/work/exec.go

```diff
@@ -51,21 +52,53 @@ func (o *Options) run() error {
 }
 
 func (o *Options) deleteWork(workClient *workclientset.Clientset) error {
-	err := workClient.WorkV1().ManifestWorks(o.Cluster).Delete(context.TODO(), o.Workname, metav1.DeleteOptions{})
-	if err != nil && errors.IsNotFound(err) {
+	work, err := workClient.WorkV1().ManifestWorks(o.Cluster).Get(context.TODO(), o.Workname, metav1.GetOptions{})
+	if err != nil {
+		if errors.IsNotFound(err) {
+			fmt.Fprintf(o.Streams.Out, "work %s is already deleted\n", o.Workname)
+			return nil
+		}
 		return err
 	}
 
-	// watch until work is fully removed
-	err = helpers.WatchUntil(
-		func() (watch.Interface, error) {
-			return workClient.WorkV1().ManifestWorks(o.Cluster).Watch(context.TODO(), metav1.ListOptions{})
-		},
-		func(event watch.Event) bool {
-			return event.Type == watch.Deleted
-		},
-	)
-	if err != nil {
+	// start a goroutine to watch the delete event
+	errChannel := make(chan error)
+	timeout := 10 * time.Second
+	go func(c chan<- error) {
+		time.Sleep(timeout)
+		c <- fmt.Errorf("delete work %s timeout, failed to delete", o.Workname)
+	}(errChannel)
+
+	go func(c chan<- error) {
+		// watch until clusterset is removed
+		e := helpers.WatchUntil(
+			func() (watch.Interface, error) {
+				return workClient.WorkV1().ManifestWorks(o.Cluster).Watch(context.TODO(), metav1.ListOptions{})
+			},
+			func(event watch.Event) bool {
+				return event.Type == watch.Deleted
+			},
+		)
+		c <- e
+
+	}(errChannel)
+
+	if o.Force {
+		work.ObjectMeta.Finalizers = work.ObjectMeta.Finalizers[:0]
+		_, err = workClient.WorkV1().ManifestWorks(o.Cluster).Update(context.TODO(), work, metav1.UpdateOptions{})
+		if err != nil {
+			return err
+		}
+	}
+
+	err = workClient.WorkV1().ManifestWorks(o.Cluster).Delete(context.TODO(), o.Workname, metav1.DeleteOptions{})
```

Comment:
run delete at first then remove finalizer, otherwise there is chance that finalizer is added back

Reply:
done